test/AWS_Tests/src/S3_Spec.enso

```diff
@@ -29,20 +30,77 @@ add_specs suite_builder =
 
         group_builder.specify "parse full paths uris" <|
-            S3.parse_uri "s3://banana/apple" . should_equal (Pair.new "banana" "apple")
-            S3.parse_uri "s3://banana/apple/orange" . should_equal (Pair.new "banana" "apple/orange")
+            S3_Path.parse "s3://banana/apple" . should_equal (S3_Path.Value "banana" "apple")
+            S3_Path.parse "s3://banana/apple/orange" . should_equal (S3_Path.Value "banana" "apple/orange")
 
         group_builder.specify "reject invalid urils" <|
-            S3.parse_uri "asda" . should_equal Nothing
-            S3.parse_uri "s3:///" . should_equal Nothing
-            S3.parse_uri "s3:///apple/orange" . should_equal Nothing
+            S3_Path.parse "asda" . should_fail_with Illegal_Argument
+            S3_Path.parse "s3:///" . should_fail_with Illegal_Argument
+            S3_Path.parse "s3:///apple/orange" . should_fail_with Illegal_Argument
+
+        group_builder.specify "should support path traversal using `/` and `parent`" <|
+            root.name . should_equal "/"
+            (root / "foo") . path . should_equal "s3://"+bucket_name+"/foo"
+            (root / "foo") . name . should_equal "foo"
+
+            (root / "foo/").is_directory . should_be_true
+            (root / "foo/" / "bar") . path . should_equal "s3://"+bucket_name+"/foo/bar"
+            # It also works if the trailing slash is missing:
+            (root / "foo" / "bar") . path . should_equal "s3://"+bucket_name+"/foo/bar"
+            # Even though it is not normally treated as directory:
+            (root / "foo").is_directory . should_be_false
+
+            (root / "foo/../././bar") . should_equal (root / "bar")
+            err = (root / "..")
+            err.should_fail_with S3_Error
+            err.catch.to_display_text . should_contain "Cannot move above root"
+
+            hello_txt.parent.parent.parent . should_equal root
+            (hello_txt / "../../..") . should_equal root
+            hello_txt.parent . should_equal (root / "examples/" / "folder 2/")
+            hello_txt.parent.is_directory . should_be_true
+
+            # Leading slash will mean starting back from bucket root:
+            (hello_txt / "/foo/bar") . should_equal (root / "foo/bar")
+            (hello_txt / "/") . should_equal root
+            (hello_txt / "////") . should_equal root
+
+        group_builder.specify "should support path traversal using `join`" <|
+            root.join ["foo", "bar"] . path . should_equal "s3://"+bucket_name+"/foo/bar"
+            root.join ["a/b/", "c/d", "e"] . path . should_equal "s3://"+bucket_name+"/a/b/c/d/e"
+            hello_txt.join ["../..", ".."] . should_equal root
+            hello_txt.join ["..", "a", ".."] . should_equal hello_txt.parent
+
+        group_builder.specify "will normalize paths upon parsing" <|
+            S3_File.new "s3://bucketA/a/b/c/../././d/../e" . path . should_equal "s3://bucketA/a/b/e"
```

Comment:
I have added a test showing that we normalize the `..`s in the `S3_File.new` constructor.

I added this, because we do such normalization in the `/` operator anyway.

IMO it would be weird if `S3_File.new "s3://a/b/../c"` denoted a file with key `b/../c` but after `S3_File.new "s3://a/b/../c" / "d"` that would now be transformed into key `c/d` because of the normalization only happening after `/`.

Reply:
However, we have to keep in mind that, while un-intuitive, it is technically possible to have an S3 file with a key `a/../b`. The `..` does not have any special meaning in S3 as far as I can tell. Moreover, the `/` is not very special either - the delimiter to use for 'virtual' directories on S3 can be customized to be any other character.

I think in the future we may want to have something like `raw_uri=True` optional argument 
 (or whatever other name, e.g. `emulate_directories=False`) which will allow the user to skip the `..` normalization and access such weird files. 

But I think we are fine for now. Worst case the user can use the lower-level `S3` methods that do not do any parsing like this. It is a trade-off between `S3_File` being more file-like and also handling such weird conventions.